Feature/branch type branch restrictions

alexhiggins732 opened the pull request:
Add support for branching model and branch permissions.

mnivet commented on SharpBucket/V2/Pocos/BranchingModelBranch.cs:
```diff
@@ -0,0 +1,9 @@
+﻿namespace SharpBucket.V2.Pocos
+{
+    public class BranchingModelBranch
```
After few tests an reading it seems to be effectively a Branch since it's the value returned by the type field by bitbucket

But since bitbucket returned only few props of the Branch type we may create a BranchInfo POCO that include just the properties returned by Bitbucket when Branch type is returned partially (with a ComitInfo type for the target property)

Using the Info suffix is the rule we try to apply everywhere in that case. Even if we still add few legacy exceptions like the UserShort and the NamedBranch types

mnivet replied:
But using the Branch type and not creating a BranchInfo could be OK
Since bitbucket seems to offer capabilities to re-add filtered properties from partial responses:
https://developer.atlassian.com/bitbucket/api/2/reference/meta/partial-response

I never tried and this is clearly not yet supported by this client, but if we go in that direction we will drop the XxxInfo types to be ale to return additional types from partial responses...

alexhiggins732 replied:
This was confusing not being familiar with the projects implementation. The `Branch` POCO simply contains the `name` property. However, looking at the other API is see that that simple POCO is used everywhere which drops all extraneous fields. As suggested I used the existing `Branch` type here.

mnivet reviewed (commented):
Tell me if you got time to work on the comments
I can plan time on my side otherwise. September was full, but October should let me more time to work on that project.

alexhiggins732 commented on SharpBucket/V2/EndPoints/RepositoryResource.cs:
```diff
@@ -203,6 +203,20 @@ public PullRequestsResource PullRequestsResource()
+        /// <summary>
+        /// Returns the branching model as applied to the repository. This view is read-only.
+        /// </summary>
+        public BranchingModel GetBranchingModel()
```
Moved to BranchingModelResource as requested using the pattern applied to tags TagsResource.

alexhiggins732 commented on SharpBucket/V2/Pocos/BranchingModel.cs:
```diff
@@ -0,0 +1,13 @@
+        public NamedBranchingModelBranch production { get; set; }
+        public List<BranchType> branch_types { get; set; }
+        public string type { get; set; }
+        public Links links { get; set; }
```
BranchingModelLinks type created.

alexhiggins732 commented on SharpBucket/V2/Pocos/NamedBranchingModelBranch.cs:
```diff
@@ -0,0 +1,9 @@
+﻿namespace SharpBucket.V2.Pocos
+{
+    public class NamedBranchingModelBranch
```
`BranchingModelBranch` was implemented.

alexhiggins732 commented on SharpBucket/V2/Pocos/Target.cs:
```diff
@@ -0,0 +1,7 @@
+﻿namespace SharpBucket.V2.Pocos
+{
+    public class Target
```
This was removed.

alexhiggins732 commented:
I addressed the issues in the comments. `RepositoryResource` now contains the additional `GetSettings` and `PutSettings` methods. These needed additional pocos created.
